timer.r/scripts/split.r

```diff
@@ -0,0 +1,176 @@
+# Assignment 1:  
+library(tweedie) 
+library(ggplot2)
+library(tictoc)
+library(knitr)
+
+
+tic()
+simTweedieTest <-  
+  function(N){ 
+    t.test( 
+      rtweedie(N, mu=10000, phi=100, power=1.9), 
+      mu=10000 
+    )$p.value 
+  } 
+toc()
+
+
+# Assignment 2:  
+
+tic()
+MTweedieTests <-  
+  function(N,M,sig){ 
+    sum(replicate(M,simTweedieTest(N)) < sig)/M 
+  } 
+toc()
+
+# Assignment 3:  
+#Load package
+library(doParallel)
+
+# Parallelized MTweedieTests
+MTweedieTests_parallel <- function(N, M, sig){
+  results <- foreach(m = 1:M, .combine = c, .packages = "tweedie") %dopar% {
+    simTweedieTest(N) < sig
+  }
+  
+  return(sum(results) / M)
+}
+
+
+# Register the number of cores to be used for parallel computation
+# This example uses 8 cores. 
+maxcores <- 8
+Cores <- min(parallel::detectCores(), maxcores)
+cl <- makeCluster(Cores)
+registerDoParallel(cl)
+
+
+
+clusterExport(cl, list("MTweedieTests_parallel", "simTweedieTest"))
+
+df <-  
+  expand.grid( 
+    N = c(10,100,1000,5000, 10000), 
+    M = 1000, 
+    share_reject = NA) 
+
+tic(paste0("Parallel loop for MTweedieTests, ", Cores, " cores"))
+for(i in 1:nrow(df)){ 
+  df$share_reject[i] <-  
+    MTweedieTests_parallel( 
+      N=df$N[i], 
+      M=df$M[i], 
+      sig=.05) 
+} 
+toc()
+
+# Close the cluster once done
+stopCluster(cl)
+
+#### 2.3 - Soulution 
+#### When splitting the m simulations in more than one core for the  MTweedieTests 
+# we get a runtime of 5.04 seconds on 8 cores. 
+# Which is drasticly faster than the 2 previos methods. 
+
+## Comparison: 
+# For 2.1 the time it takes to run the assigment 3 function is: 25.398 sec elapsed
+# for 2.2 the time is 25.079 sec 
+#for 2.3 the time is 5.04 sec
+#The reason why its faster might be simmultaneous execution. ultiple tasks can run concurrently,
+#which means multiple operations can be completed at the same time. 
+# Also Better Utilization of Resources: Modern computers have multiple cores. 
+#When tasks are run sequentially on a single core, the remaining cores remain underutilized. 
+#Parallel computing utilizes these cores, which means more tasks can be done in the same amount of time.
```

Comment:
The reason that the last one is quicker than the second one is because the second one parallelizes the rows in df. Because there are only 5 rows, having more than 5 cores will not make it quicker. The last one parallizes M, and can therefore utlize more cores in the computation